execution/executor/src/block_executor.rs

```diff
@@ -31,11 +32,92 @@ use storage_interface::{jmt_updates, DbReaderWriter};
 
 pub struct BlockExecutor<V> {
     pub db: DbReaderWriter,
+    inner: RwLock<Option<BlockExecutorInner<V>>>,
+}
+
+impl<V> BlockExecutor<V>
+where
+    V: VMExecutor,
+{
+    pub fn new(db: DbReaderWriter) -> Self {
+        Self {
+            db,
+            inner: RwLock::new(None),
+        }
+    }
+
+    pub fn root_smt(&self) -> SparseMerkleTree<StateValue> {
+        self.inner
+            .read()
+            .as_ref()
+            .expect("BlockExecutor is not reset")
+            .root_smt()
+    }
+
+    fn maybe_reset(&self) {
+        if self.inner.read().is_none() {
+            self.reset();
+        }
+    }
+}
+
+impl<V> BlockExecutorTrait for BlockExecutor<V>
+where
+    V: VMExecutor,
+{
+    fn committed_block_id(&self) -> HashValue {
+        self.maybe_reset();
+        self.inner
+            .read()
+            .as_ref()
+            .expect("BlockExecutor is not reset")
+            .committed_block_id()
+    }
+
+    fn reset(&self) {
+        *self.inner.write() = Some(BlockExecutorInner::new(self.db.clone()));
+    }
+
+    fn execute_block(
+        &self,
+        block: (HashValue, Vec<Transaction>),
+        parent_block_id: HashValue,
+    ) -> Result<StateComputeResult, Error> {
+        self.maybe_reset();
+        self.inner
+            .read()
+            .as_ref()
+            .expect("BlockExecutor is not reset")
+            .execute_block(block, parent_block_id)
+    }
+
+    fn commit_blocks_ext(
+        &self,
+        block_ids: Vec<HashValue>,
+        ledger_info_with_sigs: LedgerInfoWithSignatures,
+        save_state_snapshots: bool,
+    ) -> Result<Option<StateSnapshotDelta>, Error> {
+        self.inner
+            .read()
+            .as_ref()
+            .expect("BlockExecutor is not reset")
+            .commit_blocks_ext(block_ids, ledger_info_with_sigs, save_state_snapshots)
+    }
+
+    fn finish(&self) -> Result<()> {
+        let mut inner = self.inner.write();
+        ensure!(inner.is_some());
```

Comment:
:trollface:  can fail you when a "switch out" happens before anything is executed?

Reply:
my bad